# Homework 2
### Austin Castelo

### 1

$f(x) = e^x$

#### (a) 


In [88]:
import numpy as np

def f(x):
    return np.e**x

def D0_h(x_0,n):
    h = 0.1/(2**n)
    D = (f(x_0+h) - f(x_0-h))/(2*h)
    return D

In [89]:
D_n = []
for i in range(11):
    D_n.append(D0_h(0.5,i))
D_n

[1.6514705137461927,
 1.6494083237722656,
 1.6488930178661754,
 1.6487642064853159,
 1.6487320045835219,
 1.6487239541670462,
 1.6487219415666488,
 1.6487214384166293,
 1.6487213126291067,
 1.6487212811830432,
 1.6487212733215983]

In [90]:
for i in range(9):
    print((D_n[i]- D_n[i+1])/(D_n[i+1]- D_n[i+2]))

4.001875293014308
4.000468767990908
4.000117188221117
4.000029301751928
4.000007396336765
4.0000006354848825
3.9999994351246286
4.000103939769163
4.000036153289949


As we can see the centered difference approximation approaches a quadratic rate of convergence. The third to last term being $<4$ and the last two terms break from the convergent trend to 4 can be attributed to round-off error.

##### (b)

Since $f(x)=e^x$ is strictly increasing for $x>0$ and all of our interpolating points are $[0.4,0.6]$ we can take $M_3 = f^{'''}(0.6)$ 

In [91]:
f_x = f(0.5)
f_x

1.6487212707001282

In [92]:
M_3 = f(0.6)
eps = 2**(-52)
h_0 = (f_x*3*eps/M_3)**(1/3)
h_0

8.447159163273983e-06

In [93]:
E = 1/6*h_0**2*M_3+ f_x/h_0*eps
E

6.500818608505099e-11

In [111]:
E = 1/24*h_0**2*M_3+ f_x/(h_0/2)*eps
E

4.158645196514398e-08

In [112]:
E = 2/3*h_0**2*M_3+ f_x/(2*h_0)*eps
E

6.653676499167909e-07

As we can see reducing $h_0$ by a factor of 2 leads to a larger increase in round-off error than the decrease in approximation error, and increasing $h_0$ by a factor of 2 leads to a greater increase in approximation error than decrease in round-off error. Thus, $h_0 \approx 8.447159163273983e-06$

#### (c)

In [96]:
def R_D(x_0, n):
    R = (4*D0_h(x_0,n+1) - D0_h(x_0,n))/3
    return R

Rate of convergence should be $O(h^4)$

In [113]:
R_n = []
for i in range(11):
    R_n.append(R_D(0.5,i))
R_n

[1.64872092711429,
 1.648721249230812,
 1.6487212693583626,
 1.6487212706162573,
 1.6487212706948877,
 1.6487212706998495,
 1.648721270699956,
 1.6487212706999326,
 1.648721270701022,
 1.6487212707011167,
 1.6487212707007377]

In [114]:
for i in range(9):
    print((R_n[i]- R_n[i+1])/(R_n[i+1]- R_n[i+2]))

16.003761703618792
16.000982338385477
15.997554501298994
15.847131477669382
46.55416666666667
-4.528301886792453
-0.02160619649408887
11.489461358313818
-0.2501464557703574


In [99]:
(R_D(0.5,4)- R_D(0.5, 5))

-4.96180874165475e-12

In [100]:
(R_D(0.5,5)- R_D(0.5, 6))

-1.0658141036401503e-13

As we can see $q \approx 16$ for the first 4 terms, thus the approximation error converges at $O(h^4)$, however after the first 4 terms round-off error begins to dominate our error calculations and we lose $O(h^4)$ convergence.

Round-off error is still $O(\frac{\epsilon}{h})$ and minimium error is achieved when $O(h^4)=O(\frac{\epsilon}{h})$, therefore $h_0= \epsilon^{1/5}$

In [101]:
h_0 = eps**(1/5)
h_0

0.000740095979741405

In [102]:
(4*((f(0.5+h_0/2) - f(0.5-h_0/2))/(h_0))-((f(0.5+h_0) - f(0.5-h_0))/(2*h_0)))/3 - f(0.5)

4.107825191113079e-14

As mentioned in the notes the minimium error of $D^{ext}_h(f(x_0))$ is $ \approx 10^{-14}$ in double precision. The error at $h_0 = 0.000740095979741405$ is $4.107825191113079*10^{-14}$, thus $h_0 \approx 0.000740095979741405$.

#### (d)

Let $r=1$

In [103]:
def CTR(n):
    h = 2*np.pi/n
    T_h = (f(0.5 + 1)+ f(0.5 + np.e**(1j*2*np.pi))*np.e**-(1j*2*np.pi))/2
    for i in range(1,n):
        T_h += f(0.5 + np.e**(1j*h*i))*np.e**-(1j*h*i)
    return T_h*h

In [104]:
for i in range(1,20):
    print(f(0.5)-(CTR(i)/(2*np.pi)))

(-2.8329677996379363+0j)
(-0.28885793461258724+7.42785831001809e-17j)
(-0.06902430145698624+7.401486830834377e-17j)
(-0.013743887621111117+0j)
(-0.0022899319577727173+6.661338147750939e-17j)
(-0.000327127501018154-1.4802973661668753e-16j)
(-4.0890905791846066e-05-3.172065784643304e-17j)
(-4.543433840931144e-06+5.551115123125783e-17j)
(-4.5434338313832257e-07+1.2335811384723961e-17j)
(-4.130394404278093e-08+1.1102230246251566e-17j)
(-3.4419951333575227e-09-5.046468293750712e-17j)
(-2.6476798531405166e-10-9.25185853854297e-17j)
(-1.891176104606984e-11-4.270088556250602e-17j)
(-1.2603251775544777e-12-7.93016446160826e-18j)
(-7.838174553853605e-14+0j)
(-3.9968028886505635e-15+4.163336342344337e-17j)
(2.220446049250313e-16+1.9592171022796878e-17j)
(-6.661338147750939e-16+1.850371707708594e-17j)
(2.220446049250313e-16-3.5059674461847047e-17j)


In [105]:
f(0.5)-(CTR(17)/(2*np.pi))

(2.220446049250313e-16+1.9592171022796878e-17j)

As we can see we reach machine precision with the composite trapezodial rule applied to Cauchy's Integral Formula with 18 equidistant interpolating nodes.


### 2)

$$D_h(f(x_0)) = \frac{-f(x_0+2h)+4f(x_0+h)-3f(x_0)}{2h}$$

$\textbf{Proof:}$

$ f^{'}(x_0)-D_h(f(x_0)) = \frac{1}{3}h^2f^{'''}(\epsilon) \iff f^{'}(x_0) - \frac{-f(x_0+2h)+4f(x_0+h)-3f(x_0)}{2h} = \frac{1}{3}h^2f^{'''}(\epsilon) $

Taylor Series expansion

$f(x_0+h)= f(x_0)+f^{'}(x_0)h+\frac{1}{2}f^{''}(x_0)h^2+\frac{1}{3!}f^{(3)}(\epsilon_1)h^3$

$f(x_0+2h)= f(x_0)+f^{'}(x_0)2h+\frac{1}{2}f^{''}(x_0)4h^2+\frac{1}{3!}f^{(3)}(\epsilon_2)8h^3 = f(x_0)+2f^{'}(x_0)h+2f^{''}(x_0)h^2+\frac{4}{3}f^{(3)}(\epsilon_2)h^3$

Where $x_0 \leq \epsilon_1 \leq x_0+h$ and $x_0 \leq \epsilon_2 \leq x_0+2h$

Choose $\epsilon = max{(\epsilon_1,\epsilon_2)}$

$f^{'}(x_0) - \frac{-f(x_0+2h)+4f(x_0+h)-3f(x_0)}{2h} = \frac{1}{3}h^2f^{'''}(\epsilon) \iff f^{'}(x_0)- \frac{-(f(x_0)+2f^{'}(x_0)h+2f^{''}(x_0)h^2+\frac{4}{3}f^{(3)}(\epsilon)h^3)+4(f(x_0)+f^{'}(x_0)h+\frac{1}{2}f^{''}(x_0)h^2+\frac{1}{3!}f^{(3)}(\epsilon)h^3)-3f(x-0)}{2h} = \frac{1}{3}h^2f^{'''}(\epsilon)$

$\iff f^{'}(x_0)- \frac{3f(x_0)+2f^{'}(x_0)h-\frac{2}{3}f^{(3)}(\epsilon)h^3-3f(x_0)}{2h} =  \frac{1}{3}h^2f^{'''}(\epsilon) $

$\iff f^{'}(x_0)-\frac{2hf^{'}(x_0)-\frac{2h^3}{3}f^{(3)}(\epsilon)}{2h} = \frac{1}{3}h^2f^{'''}(\epsilon)$

$\iff f^{'}(x_0)-f^{'}(x_0)-\frac{h^2}{3}f^{(3)}(\epsilon)= \frac{1}{3}h^2f^{'''}(\epsilon)$

$\iff \frac{1}{3}h^2f^{(3)}(\epsilon)= \frac{1}{3}h^2f^{'''}(\epsilon)$

### 3)

#### (a)

$\textbf{Proof:}$

$l_j(x)=\frac{1}{\alpha_j} \prod^{n}_{k=0 k\neq j} (x-x_k)$

$\iff \alpha_j l_j(x)=\prod^{n}_{k=0 k\neq j} (x-x_k)$

$\iff ln(\alpha_j l_j(x))= ln(\prod^{n}_{k=0 k\neq j} (x-x_k))= \sum^{n}_{k=0 k\neq j} ln(x-x_k)$

Take derivative of left-hand side and right-hand side


$\frac{d}{dx} ln(\alpha_j l_j(x))=\frac{\frac{d}{dx}(\alpha_j l_j(x))}{\alpha_j l_j(x)}=\frac{\alpha_j}{\alpha_j}\frac{1}{l_j(x)}l_j^{'}(x)= \frac{l_j^{'}(x)}{l_j(x)}$

$\frac{d}{dx} \sum^{n}_{k=0 k\neq j} ln(x-x_k)=\sum^{n}_{k=0 k\neq j} \frac{1}{x-x_k}$

Then,

$\frac{l_j^{'}(x)}{l_j(x)}=\sum^{n}_{k=0 k\neq j} \frac{1}{x-x_k} \iff l^{'}_j(x)= l_j(x)[\sum^{n}_{k=0 k\neq j}\frac{1}{x-x_k}]$

#### (b)

$\textbf{Proof 1:} (D_n)_{ij}=\frac{\alpha_i}{\alpha_j}\frac{1}{x_i-x_j} $

$(D_n)_{ij}=l^{'}_j(x_i)$

Assume $i \neq j$ and $l^{'}_j(x_i)= l_j(x_i)[\sum^{n}_{k=0 k\neq j}\frac{1}{x_i-x_k}]$

$l^{'}_j(x_i)= \sum^{n}_{k=0 k\neq j} \frac{\frac{1}{\alpha_j} \prod^{n}_{m=0 m\neq j}(x_i-x_m)}{x_i-x_k}= \sum^{n}_{k=0 k\neq j} [{\frac{1}{\alpha_j} \prod^{n}_{m=0 m\neq j,k}(x_i-x_m)}\frac{x_i-x_k}{x_i-x_k}]$

for $x_i \neq x_k$ for some $k \in \{1,2,...,n\}-\{j\}$

$l^{'}_j(x_i)= \sum^{n}_{k=0, k\neq j} \frac{1}{\alpha_j}\prod^{n}_{m=0,m\neq j,k}(x_i-x_m) = \frac{1}{\alpha_j}\sum^{n}_{k=0, k\neq j} \prod^{n}_{m=0,m\neq j,k}(x_i-x_m)$

$\forall k \neq i \prod^{n}_{m=0,m\neq j,k}(x_i-x_m)=0$

Then, $\frac{1}{\alpha_j}\prod^{n}_{m=0,m\neq j,k}(x_i-x_m)= \frac{1}{\alpha_j}\alpha_i\frac{1}{x_i-x_j}$

$l^{'}_j=\frac{\alpha_i}{\alpha_j}\frac{1}{x_i-x_j} = (D_n)_{ij}$

$\textbf{Proof 2:} (D_n)_{ii}=\sum^{n}_{k=0,k\neq i}\frac{1}{x_i-x_k}$

$(D_n)_{ii}=l^{'}_i(x_i) = l_i(x_i)[\sum^{n}_{k=0 k\neq i}\frac{1}{x_i-x_k}]$

$l_i(x_i)= \prod^{n}_{k=0, k\neq j} \frac{x_i-x_k}{x_i-x_k}= \prod^{n}_{k=0, k\neq j} 1$

$l^{'}_i(x_i)= \sum^{n}_{k=0 k\neq i}\frac{1}{x_i-x_k}$

$(D_n)_{ii}=\sum^{n}_{k=0 k\neq i}\frac{1}{x_i-x_k}$


#### (c)

$\sum^{n}_{j=0}(D_n)_{ij}=\sum^{n}_{j=0}\frac{\alpha_i}{\alpha_j}\frac{1}{x_i-x_j}+(D_n)_{ii}$

Take i=0

$\sum^{n}_{j=0}(D_n)_{0j}=\sum^{n}_{j=0}\frac{\alpha_0}{\alpha_j}\frac{1}{x_0-x_j}+\sum^n_{k=1}\frac{1}{x_0-x_k} =\frac{\alpha_0}{\alpha_1}\frac{1}{x_0-x_1}+\frac{\alpha_0}{\alpha_2}\frac{1}{x_0-x_2}+...+\frac{\alpha_0}{\alpha_n}\frac{1}{x_0-x_n}$

$\frac{(x_0-x_1)(x_0-x_2)...(x_0-x_n)}{(x_1-x_0)(x_1-x_2)...(x_1-x_n)}\frac{1}{(x_0-x_1)} + \frac{(x_0-x_1)(x_0-x_2)...(x_0-x_n)}{(x_2-x_0)(x_2-x_1)...(x_2-x_n)}\frac{1}{(x_0-x_2)}...+\frac{(x_0-x_1)(x_0-x_2)...(x_0-x_n)}{(x_n-x_0)(x_n-x_2)...(x_n-x_{n-1})}\frac{1}{(x_0-x_n)} +\frac{1}{x_0-x_1}+\frac{1}{x_0-x_2}+...+\frac{1}{x_0-x_n}$

$= \frac{-1}{x_0-x_1}+\frac{-1}{x_0-x_2}+...+\frac{-1}{x_0-x_n}+\frac{1}{x_0-x_1}+\frac{1}{x_0-x_2}+\frac{1}{x_0-x_n} = 0$

$\rightarrow \sum^{n}_{j=0}(D_n)_{0j}=0$

Similiarly for other i's

$\frac{(x_i-x_1)(x_i-x_2)...(x_i-x_n)}{(x_1-x_0)(x_1-x_2)...(x_1-x_n)}\frac{1}{(x_i-x_1)} + \frac{(x_i-x_1)(x_i-x_2)...(x_i-x_n)}{(x_2-x_0)(x_2-x_1)...(x_2-x_n)}\frac{1}{(x_i-x_2)}...+\frac{(x_i-x_1)(x_i-x_2)...(x_i-x_n)}{(x_n-x_0)(x_n-x_2)...(x_n-x_{n-1})}\frac{1}{(x_i-x_n)} +\frac{1}{x_i-x_1}+\frac{1}{x_i-x_2}+...+\frac{1}{x_i-x_n}$

$= \frac{-1}{x_i-x_1}+\frac{-1}{x_i-x_2}+...+\frac{-1}{x_i-x_n}+\frac{1}{x_i-x_1}+\frac{1}{x_i-x_2}+\frac{1}{x_i-x_n} = 0$

$\rightarrow \sum^{n}_{j=0}(D_n)_{ij}=0 \forall i=0,1,...,n$

#### (d)

$x=[-1,0,1]$

$D_{0,0}= \sum^{2}_{k=1} \frac{1}{x_i-x_k} = \frac{1}{x_0-x_1}+\frac{1}{x_0-x_2}= \frac{1}{-1-0}+\frac{1}{-1-1}= -1+\frac{1}{-2}= \frac{-3}{2}$

$D_{1,1}=\frac{1}{0-(-1)}+\frac{1}{0-1}= 1+ -1=0$

$D_{2,2}=\frac{1}{1-(-1)}+\frac{1}{1-0}= \frac{3}{2}$

$\alpha_0 = (x_0-x_1)(x_0-x_2)=(-1-0)(-1-1)=2$

$\alpha_1 = (x_1-x_2)(x_1-x_0)=(0-1)(0-(-1))=-1$

$\alpha_2 = (x_2-x_1)(x_2-x_0)=(1-0)(1-(-1))=2$

$D_{01}=\frac{\alpha_0}{\alpha_1}(x_0-x_1)=\frac{2}{-1}(-1-0)=2$

$D_{10}=\frac{\alpha_1}{\alpha_0}(x_1-x_0)=\frac{-1}{2}(0-(-1))=\frac{-1}{2}$

$D_{21}=\frac{\alpha_2}{\alpha_1}(x_2-x_1)=\frac{2}{-1}(1-0)=-2$

$D_{12}=\frac{\alpha_1}{\alpha_2}(x_1-x_2)=\frac{-1}{2}(0-1)=\frac{1}{2}$

$D_{02}=\frac{\alpha_0}{\alpha_2}(x_0-x_2)=\frac{2}{2}(-1-1)=-2$

$D_{20}=\frac{\alpha_2}{\alpha_0}(x_2-x_0)=\frac{2}{2}(1-(-1))=2$

$
D_2=
  \begin{bmatrix}
    \frac{-3}{2} & \frac{-1}{2} & 2 \\
    2 & 0 & -2 \\
    -2 & \frac{1}{2} & \frac{3}{2}
  \end{bmatrix}
$